vql/tools/starlark.go

```diff
@@ -0,0 +1,522 @@
+				return nil, fmt.Errorf("%w: %v", ErrStarlarkConversion, err)
+			}
+		}
+
+		return result, nil
+	case map[string]interface{}:
+		result := map[string]starlark.Value{}
+
+		for key, value := range v {
+			mapValueStarlarked, err := interfaceAsStarlarkValue(value)
+			if err != nil {
+				return nil, err
+			}
+
+			result[key] = mapValueStarlarked
+		}
+
+		dict := starlarkstruct.FromStringDict(starlarkstruct.Default, result)
+
+		return dict, nil
+	case map[int32]interface{}:
+
+		keyword_tuple := make([]starlark.Tuple, 0)
+		for key, value := range v {
+			mapValueStarlarked, err := interfaceAsStarlarkValue(value)
+			if err != nil {
+				return nil, err
+			}
+			new_val := int(key)
+			keyword_tuple = append(keyword_tuple, starlark.Tuple{starlark.MakeInt(new_val), mapValueStarlarked})
+		}
+
+		dict := starlarkstruct.FromKeywords(starlarkstruct.Default, keyword_tuple)
+
+		return dict, nil
+
+	case *ordereddict.Dict:
+		// Most of the time we're here
+		result := map[string]starlark.Value{}
+
+		// Convert data to JSON and back to make sure it's types starlark can understand
+		data, err := v.MarshalJSON()
+		if err != nil {
+			return nil, err
+		}
+		new_dict := ordereddict.NewDict()
+		err = new_dict.UnmarshalJSON(data)
+		if err != nil {
+			return nil, err
+		}
+		for key, value := range *new_dict.ToDict() {
+			mapValueStarlarked, err := interfaceAsStarlarkValue(value)
+			if err != nil {
+				return nil, err
+			}
+
+			result[key] = mapValueStarlarked
+		}
+
+		dict := starlarkstruct.FromStringDict(starlarkstruct.Default, result)
+
+		return dict, nil
+
+	default:
+		return nil, fmt.Errorf("%w: unsupported type %T", ErrStarlarkConversion, value)
+	}
+}
+
+func tryIntegerOrFloat(floatValue float64) (starlark.Value, error) {
+	if integerValue := int64(floatValue); floatValue == float64(integerValue) {
+		return starlark.MakeInt64(integerValue), nil
+	}
+
+	return starlark.Float(floatValue), nil
+}
+
+// Get scope of compiled starlark context
+func getCompiled(ctx context.Context,
+	scope vfilter.Scope,
+	key string, code string, globals vfilter.Any) (*ordereddict.Dict, error) {
+	if key == "" {
+		new_hash := sha256.New()
+		new_hash.Write([]byte(code))
+		key = string(new_hash.Sum(nil))
+	}
+
+	compiled_vars, ok := vql_subsystem.CacheGet(scope, key).(*ordereddict.Dict)
+	if !ok {
+		compiled_vars = ordereddict.NewDict()
+		err := compileStarlark(code, globals, compiled_vars)
+		if err != nil {
+			return nil, err
+		}
+		vql_subsystem.CacheSet(scope, key, compiled_vars)
+	}
+
+	return compiled_vars, nil
+}
+
+// User doesn't actually specify extras, that's done in extractStarlarkFuncArgs
+type StarlarkFuncArgs struct {
+	Func    string      `vfilter:"required,field=func,doc=Starlark function to call."`
+	Code    string      `vfilter:"required,field=code,doc=Starlark code to compile."`
+	Args    vfilter.Any `vfilter:"optional,field=args,doc=Positional args for the function."`
```

Comment:
You probably want a LazyExpr here. The only time you need an Any is if you want to do the materializing yourself. Any literally does not do any conversion or type checking so you have to do a lot of it by hand.

A LazyExpr will be available when you call Reduce()